Add New Code Sample: memory_system (#3)
haoyanwa/oneAPI-samples: DirectProgramming/C++SYCL_FPGA/Tutorials/DesignPatterns/memory_system/README.md

> # `Memory System` Sample

Reviewer: I think `Banked Memory System` would be a more appropriate title

You should replace
~`memory_system`~ -> `banked_memory_system`
in the rest of this sample

> This code sample demonstrates a design pattern in optimizing banked memory system.

Reviewer: ```suggestion
This code sample demonstrates how to optimize a banked memory system.
```

## Purpose

> In comparison to registers, memory gives you the flexibility of multiple read and write accesses. However, an ill-formed memory layout can lead to negative impact on your system performance. 

Reviewer: ```suggestion
When compiling your design for FPGA architectures, the Intel® oneAPI DPC++/C++ Compiler can choose to implement your variables as registers, or as block memories. The details of this decision are better defined in the [Memory System code sample](DirectProgramming/C++SYCL_FPGA/Tutorials/Features/memory_attributes) However, an ill-formed memory layout can lead to negative impact on your system performance. 
```

> In this code sample, we will see:

Reviewer: ```suggestion
In this code sample, you will learn:
```

> In this code sample, we will see:

Reviewer: these items should be listed in the order they appear in the text below.:
* When to use a banked memory and how the compiler optimizes banked memory systems
* how to recognize stallable arbitration...
* how to write code that results in efficient banked memory systems

> * What stallable arbitration looks like and how it leads to bad system performance.

Reviewer: ```suggestion
* How to recognize stallable arbitration in the FPGA optimization report, and how it can lead to bad system performance.
```

> * How to leverage banked memory system in the correct way.

Reviewer: ```suggestion
* How and when to construct a banked memory system to improve throughput in your designs.
```

| Optimized for      | Description
|:---                |:---
| OS                 | Ubuntu* 20.04 <br> RHEL*/CentOS* 8 <br> SUSE* 15 <br> Windows* 10 <br> Windows Server* 2019
| Hardware           | Intel® Agilex® 7, Agilex® 5, Arria® 10, Stratix® 10, and Cyclone® V FPGAs
| Software           | Intel® oneAPI DPC++/C++ Compiler

> **Note**: Even though the Intel DPC++/C++ oneAPI compiler is enough to compile for emulation, generating reports and generating RTL, there are extra software requirements for the simulation flow and FPGA compiles.
>
> For using the simulator flow, Intel® Quartus® Prime Pro Edition (or Standard Edition when targeting Cyclone® V) and one of the following simulators must be installed and accessible through your PATH:
> - Questa*-Intel® FPGA Edition
> - Questa*-Intel® FPGA Starter Edition
> - ModelSim® SE
>
> When using the hardware compile flow, Intel® Quartus® Prime Pro Edition (or Standard Edition when targeting Cyclone® V) must be installed and accessible through your PATH.
>
> :warning: Make sure you add the device files associated with the FPGA that you are targeting to your Intel® Quartus® Prime installation.

## Key Implementation Details

> In this code sample, we have two kernels `NaiveKernel` and `OptimizedKernel`.

Reviewer: ```suggestion
```
no point mentioning this yet

The sample design shows access to a buffer implemented in on-chip memory. Read and write operations relies on LSUs.

> This tutorial simplifies a real-world access pattern to the banked memory system. There is a `buffer` that is implemented in on-chip memory, which accepts read and write. We write to the buffer and read from the buffer in a loop, which we expect an initiation interval of 1. This means that the read and write to the memory system need to be pipelined. Therefore, a number of read and write to the memory system in the loop body should all happen in 1 clock cycle.

Reviewer: ```suggestion
The sample design shows access to a 2-dimensional FIFO buffer implemented in on-chip memory. The design performs a read-modify-write on the buffer in the `col` loop, which should have an initiation interval of 1. The fully-unrolled `row` loop accesses an entire dimension of the buffer in **the same clock cycle**.
```

Pipelining the reads and writes is irrelevant; the important thing is that the loop iterations are pipelined and that the read/modify/write operations on each row have to happen at the **same time**.

> Each memory block (in this case MLAB) has a read port and a write port. For true dual-port memory setup, we can issue one read and one write in a single clock cycle. As a result, when the number of access exceeds the capacity of a memory block, the memory will become the bottleneck. Consequently, efficient access to the memory system becomes a key to the system performance.

Reviewer: ```suggestion
Each memory block has a read port and a write port, so the design can issue one read and one write in the same clock cycle. If the design tries to perform more than 2 accesses in the same clock cycle, arbitration hardware is inserted to resolve any possible access conflicts. If this arbitration hardware has to be used, your design's throughput will suffer. To guarantee the maximum possible throughput, try to avoid arbitration logic in your memory systems.
```

in your source code, you asked for 'block RAM' not 'MLAB', so saying 'MLAB' here is probably not correct. If the compiler is giving you MLABs anyway, then the compiler is ignoring your directive.

Author: I must have glimpsed on the wrong thing. I can confirm the report says block RAM.
![image](https://github.com/user-attachments/assets/b1923483-be07-4df9-bc28-65a4c0cc0fc1)

> 

Reviewer: ```suggestion
struct NaiveKernel {
  void operator()() const {    
    [[intel::fpga_memory("BLOCK_RAM")]]
    int buffer[kNumRows][kNumCols];

    [[intel::initiation_interval(1)]]
    for (int col = 0; col < kNumCols; ++col) {
      SimpleInputT input = InStream_NaiveKernel::read();

      #pragma unroll
      for (int row = 0; row < kNumRows - 1; ++row) {
        buffer[row][col] = buffer[row + 1][col];
      }
      buffer[kNumRows - 1][col] = input;

      SimpleOutputT output;

      #pragma unroll
      for (int idx = 0; idx < kNumRows; ++idx) {
        output[idx] = buffer[idx][col];
      }

      OutStream_NaiveKernel::write(output);
    }
  }
};
```
You should include the kernel code at this point in addition to describing it, so the reader understands what you're describing

> 

Reviewer: Move the animation up here; the animation shows the **desired** behaviour of the kernel so it belongs up here.

### Naive Kernel

> The access pattern in the `NaiveKernel` is illustrated in the following diagram:

Reviewer: ```suggestion
The desired access pattern in the `NaiveKernel` is illustrated in the following diagram:
```

> 

Reviewer: nice graphic! 2 comments on how to make it a bit clearer (I had to watch it several times):

1. Do the shifts first, then slide `input` in a half-second or so later. This follows the sequence in your code.
2. don't forget the 'output' step, when you concurrently read from all banks. The compiler _may_ add a replicate (viewable in the memory viewer) in this case, but it may also be smart enough to re-use the value from the read that you did in the 'shift' loop.

![Naive Kernel Memory Access Pattern](./assets/access_pattern_naive.gif)

> In the `NaiveKernel`, the first array dimension (row) is unrolled in a loop. In this way, the kernel tries to access all five elements in one column of the buffer, namely `buffer[0][col]` to `buffer[4][col]`. As a result, there are 4 loads and 5 stores to the memory system in each loop iteration, and since the initiation interval is 1, they happens in a single clocl cycle. The memory layout for the `buffer` in this case has 8 banks. Looking at the Details from the Kernel Memory Viewer, we can learn that the bank bits are second to the fourth least significant bits $b_4$, $b_3$, $b_2$. This essentially means that integers in every other column will be in a different memory bank.

Reviewer: ```suggestion
In the `NaiveKernel`, the first array dimension (row) is unrolled in a loop. In this way, the kernel tries to access all five elements in one column of the buffer, namely `buffer[0][col]` to `buffer[4][col]`. As a result, there are 4 loads and 5 stores to the memory system in each loop iteration, and since the initiation interval is 1, they happen in a single clock cycle. The  [kernel memory viewer report](https://www.intel.com/content/www/us/en/docs/oneapi-fpga-add-on/developer-guide/current/kernel-memory-viewer.html) shows that the memory system for `buffer` has 8 banks. 
<move kernel memory viewer screenshot here>

#### Making sense of the bank-selection bits
The 'Details' pane of the kernel memory viewer report shows that the bank bits are the second to the fourth least significant bits, $b_4$, $b_3$, $b_2$. This means that integers in every other column will be in a different memory bank.
```

what about the least-significant bits $b_1$ and $b_0$? I suppose they are there because there are 4 bytes in each word?? You should confirm with someone (and/or an experiment with a wider or narrower type :))

Author: Yes, I checked the documentation that the last two bank bits are for sub-word access. Screenshot is taken from the OpenCL best practice guide. I couldn't find this in the Handbook, which I feel that it should be added. If you think so, I will add it in the next review.
![image](https://github.com/user-attachments/assets/fcf25a14-1f7b-42b5-9fe2-1e5fc6028ac1)

Also in the details:
![image](https://github.com/user-attachments/assets/867ca5f5-93ad-45e2-b38b-885fcae463d1)

Reviewer: You can open an HSD ES to tech comms by following these instructions:
https://wiki.ith.intel.com/pages/viewpage.action?pageId=1105823292
make sure Joe Garvey reviews it

> 

Reviewer: ```suggestion
The following table shows how each address encodes its bank of the memory system. Note how the banks do not align with the rows.
```

|                | Address | Bank bits | Bank # |
|----------------|---------|-----------|--------|
| buffer[0][0]   | 0x0000  | 0b000     | 0      |
| buffer[0][1]   | 0x0004  | 0b001     | 1      |
| buffer[0][2]   | 0x0008  | 0b010     | 2      |
| buffer[0][3]   | 0x000C  | 0b011     | 3      |

> | buffer[0][4]   | 0x0010  | 0b100     | 4      |

Reviewer: show bank 0, 1, 2, ... 7, 0 , 1, 2, ... 7 instead of 0, 1, 2, 3, 4, ... 3, 4 to make it clear that the bank demarkations happen at different points than the row ends. It shouldn't matter, because here you show the row ends don't match up with bank transitions, but in your text you drew attention to the bank bits.

| ...            |         |           |        |
| buffer[0][499] | 0x07CC  | 0b011     | 3      |
| buffer[1][0]   | 0x07D0  | 0b100     | 4      |
| ...            |         |           |        |

> We can produce this table showing which integer is stored in which bank of the memory system. 

Reviewer: ```suggestion
Similarly, the following table shows which memory bank each logical array element gets assigned to. Observe that each access to one column of `buffer` creates 2 stores and 3 loads to even numbered memory banks and 2 stores and 2 loads to odd numbered memory banks. Since more than one load and store is scheduled to the same memory bank in the same clock cycle, the compiler inserts arbitration logic along with stallable load-store units (LSUs). You can see this in the Kernel Memory Viewer in the optimization report.
```

If you are referring to a named variable, such as `buffer`, you should not use 'the' 😊

> ![Naive Kernel Memory Layout](./assets/memory_system_naive.png)

Reviewer: nice graphic. I absolutely see what you are trying to do and i think it really clearly communicates why there is so much arbitration. I recommend two changes to make it more clear:
1. The stripes are incredibly misleading; please color-code the banks so that they match the bank numbers
2. add an overlay to show what array elements are accessed concurrently.

> Observe that each access to one column of the `buffer` creates 2 stores and 3 loads to even numbered memory banks and 2 stores and 2 loads to odd numbered memory banks. These number of load and store surpasses the physical capacity of the memory blocks, and therefore, the compiler inferrs stallable LSUs as arbitration node in the Kernel Memory Viewer.

Reviewer: ```suggestion
```
you nailed me for this in a previous code sample: the descriptive text should come before the picture, so I know what I'm looking at 😊

![Stallable Arbitration Node](./assets/memory_report_naive.png)

> The stallable arbitration inside the loop body will stall the downstream logic from making progress, and in this example, we expect stall in pipe write for the `OutStream_NaiveKernel` pipe of which it's output depends on the current column of memory that is being modified. We can verify this behaviour in the simulation waveform: there are constant "dips" in the valid signal for pipe write - i.e. the valid signal is desserted because of the upstream stall from the memory system. This can negatively affect the system throughput and cause outer loop to have serial execution.

Reviewer: ```suggestion
When the kernel executes and simultaneous accesses to the memory system get arbitrated, the loop stalls. These stalls are visible in the simulation waveform: there are occasional gaps in the `valid` signal for pipe write - the valid signal is de-asserted because of the upstream stall from the memory system. This negatively affects the system throughput.
```

~behaviour~ behavior -> As much as it pains me, we must use American English.

![Naive System Waveform](./assets/waveform_naive.png)

### Optimized Kernel

To make the memory system more efficient, we can make three changes:

> 1. We use the right-most array dimension for unrolled accesses, i.e. column dimension.

Reviewer: ```suggestion
1. Use the right-most array dimension for unrolled accesses, (that is, store the *column* dimension in the right-most array dimension).
```

> 2. We choose the number of banks to a power-of-2 using `constexpr`-math functions. This line of code computes the closest power-of-2 above the given number. We use this power-of-2 as the number of banks in our optimized memory system. In our case, we use 8.

Reviewer: ```suggestion
2. Choose the number of banks to a power-of-2 using `constexpr`-math functions. This line of code computes the closest power-of-2 above the given number. We use this power-of-2 as the number of banks in our optimized memory system. Since this design performs up to 5 parallel memory accesses, a minimum of 5 banks are required. Since the next power-of-2 is 8, this design should be specified with 8 banks.
```

   > ```
   > constexpr size_t kNumBanks = fpga_tools::Pow2(fpga_tools::CeilLog2(kNumRows));
   > ```

> 3. We change the loop induction variables to match with the memory access pattern and result in the correct indices to the `buffer`.

Reviewer: ```suggestion
3. Change the loop induction variables to match with the memory access pattern and result in the correct indices to the `buffer` as assigned in change 1.
```

After making the changes, we have the `OptimizedKernel` exercising the following memory access pattern:

> ![Optimized Kernel Memory Access Pattern](./assets/access_pattern_optimized.gif)

Reviewer: the stripes should be consistent with the first graphic; it looks like the stripes are supposed to indicate banks so since you have the memory system in column-major order now, the stripes should be horizontal. 
Also you should gray-out the 3 'padding' banks to make it clear that you expect them to be optimized away.

Note that in the Kernel Memory Viewer, although the same bank bits ($b_4$, $b_3$, $b_2$) are used, this time, the memory partitioning is exactly what desired. We can deduce a memory layout using the same table:

|              | Address | Bank bits | Bank # |
|--------------|---------|-----------|--------|
| buffer[0][0] | 0x0000  | 0b000     | 0      |

> | buffer[0][1] | 0x0004  | 0b001     | 1      |

Reviewer: partition this table like how I suggested in the previous section

| buffer[0][2] | 0x0008  | 0b010     | 2      |
| buffer[0][3] | 0x000C  | 0b011     | 3      |
| buffer[0][4] | 0x0010  | 0b100     | 4      |
| buffer[0][5] | 0x0014  | 0b010     | 5      |
| buffer[0][6] | 0x0018  | 0b011     | 6      |
| buffer[0][7] | 0x001C  | 0b100     | 7      |
| buffer[1][0] | 0x0020  | 0b000     | 0      |
| buffer[1][1] | 0x0024  | 0b001     | 1      |
| ...          |         |           |        |

Because that we are only accessing the first 5 columns of the `buffer`, that provides sufficient knowledge for the compiler to optimized out the unused memory banks - bank 5 to 7. As a result, we have the following memory layout:

> ![Optimized Kernel Memory Layout](./assets/memory_system_optimized.png)

Reviewer: this graphic is probably redundant

Author: I would like to kept it for consistency and since we updated the color-coding, it makes more sense now :)

Review the memory access pattern above, we can easily find out that within each loop iteration, each of the memory banks is access precisely twice: 1 load and 1 store. This time, the memory accesses do not surpass the physical limit of the memory block hardware, and thus, we have eliminated the memory bottleneck. In the meantime, observe that the Kernel Memory Viewer shows pipelined never-stall LSUs for our optimized memory system, and the compiler successfully optimized out the unused memory banks.

![Never-stall Node](./assets/memory_report_optimized.png)

Subsequently, since we have made the LSU non-stallable, we expect no stalling in the downstream task - that the output pipe `OutStream_OptimizedKernel` is not stalled by the memory access in the loop. To verify this improvement, refer to the simulation waveform and observe the `valid` and `data` signal of the output pipe `ID_OutStream_OptimizedKernel_pipe_channel_write`. The "dips" will no longer appear in the `valid` signal, and the pipe writes out data every clock cycle.

> ![Optimized System Waveform](./assets/waveform_optimized.png)

Reviewer: this screenshot is a bit too short; please make it narrower so that it fills up more vertical space and is easier to read.

### On a Windows* System
This design uses CMake to generate a build script for  `nmake`.

1. Change to the sample directory.

2. Configure the build system for the Agilex® 7 device family, which is the default.
   ```
   mkdir build
   cd build
   cmake -G "NMake Makefiles" ..
   ```

   > **Note**: You can change the default target by using the command:
   > ```
   > cmake -G "NMake Makefiles" .. -DFPGA_DEVICE=<FPGA device family or FPGA part number>
   > ```

3. Compile the design through the generated `Makefile`. The following build targets are provided, matching the recommended development flow:

   | Compilation Type    | Command (Windows)
   |:---                 |:---
   | FPGA Emulator       | `nmake fpga_emu`
   | Optimization Report | `nmake report`
   | FPGA Simulator      | `nmake fpga_sim`
   | FPGA Hardware       | `nmake fpga`

   > **Note**: If you encounter any issues with long paths when compiling under Windows*, you may have to create your 'build' directory in a shorter path, for example c:\samples\build.  You can then run cmake from that directory, and provide cmake with the full path to your sample directory, for example:

>       > ```
>          > cmake -G "NMake Makefiles" C:\long\path\to\code\sample\CMakeLists.txt
>       > ```

Reviewer: ```suggestion
      > ```
      > > cmake -G "NMake Makefiles" C:\long\path\to\code\sample\CMakeLists.txt
      > ```
```

this renders weird
![image](https://github.com/user-attachments/assets/16900e74-8341-418b-84b6-17b54eac8524)

Author: They look pretty good in VSCode tho. Probably the nuance between rendering engines.
![image](https://github.com/user-attachments/assets/c7df3636-6222-4a49-b0a6-d9d84488633e)

Author: fixed.
